Datahub integration fetch metadata

saruniitr opened the pull request:
## Change description

This PR adds changes to enable Data Hub integration in the tool.
It add `ACCESS_TOKEN`, `HAWK` authentication related keys to the environment which enables the tool to use Data Hub services and fetch metadata.

Helper functions are added to do Company search, contact search from the tool. These are useful at the time of Investment enquiry creation to see if a company/contact already exists.

Other functions added to fetch uuids for various investment related fields. The reference data in the tool are mapped to corresponding uuid before they are sent to Data Hub.

These changes are mainly in preparation for the another PR in which the actual investment enquiry will be posted. This is split out to keep the size of PR shorter, if everything is preferred in a single PR please let me know, I will update this accordingly.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,182 @@
+    else:
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {settings.DATA_HUB_ACCESS_TOKEN}",
```
The access token for the OAuth endpoints comes from Staff SSO – it isn't an application one, it identifies a user.

saruniitr replied:
Yes agree, SSO integration hasn't happened yet so for me to test I am sending it in the headers.

reupen replied:
If that's the case, please add a TODO here and in the configuration file, so it doesn't cause any confusion later on (for other people).

saruniitr replied:
ok

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,182 @@
+
+    try:
+        response = requests.request(
+            method, url, headers=headers, data=payload, timeout=timeout
```
Would suggest passing `json=<data structure>` rather than `data=json.dumps(...)`.

saruniitr replied:
ok, agree it is better

reupen commented on app/enquiries/tests/test_dh_integration.py:
```diff
@@ -0,0 +1,23 @@
+        payload = {"name": "test"}
+
+        with pytest.raises(Timeout):
+            response = dh_request("POST", url, payload, timeout=2)
```
Would consider using `requests-mock` rather than patching `requests` directly – it will give more realistic behaviour: https://requests-mock.readthedocs.io/en/latest/

saruniitr replied:
Couldn't get it to work, will revisit later but for now will keep the current approach.

reupen commented on app/enquiries/tasks.py:
```diff
@@ -0,0 +1,31 @@
+            rds.set(key, json.dumps(dh_metadata))
+            logging.info(f"Data Hub metadata last refreshed at {datetime.now()}")
+        except ConnectionError:
+            logging.error("Error connecting to Redis backend, cannot fetch metadata")
```
I would consider using the Django cache framework rather than using Redis directly: 

- https://docs.djangoproject.com/en/3.0/topics/cache/
- http://jazzband.github.io/django-redis/latest/

saruniitr replied:
definitely simplifies, modified.

saruniitr commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,182 @@
+from rest_framework import status
+
+
+def dh_request(method, url, payload, request_headers=None, timeout=15):
```
This is a preparatory PR adding mostly helper functions accessing Data hub API. The next PR uses these functions to create investments, my plan is to include more tests in that PR.

Also I really do not want to add tests that test Data hub API because they already exist hence added a test that simulates failure.
Besides unit testing these functions any suggestions on what tests you would like to see?

reupen replied:
There is quite a lot of logic in here about caching, error handling and transforming results. I really think some tests are warranted.

saruniitr replied:
I have added tests to cover caching and connection issues. I am going add few more cases to cover company and contact search, hopefully that is sufficient for this PR.
As I mentioned earlier, this is preparation work and the next PR actually uses these functions. I will extend these tests in that PR as that gives complete picture.

saruniitr commented on app/enquiries/tasks.py:
```diff
@@ -0,0 +1,31 @@
+            rds.ping()
+
+            dh_metadata = dh_fetch_metadata()
+            rds.set(key, json.dumps(dh_metadata))
```
Yes that is the plan. I was not keen to couple caching changes in that function but since it needs to cached in some way it makes sense to handle caching within that function. I will modify this.

saruniitr replied:
updated function to handle caching also.

saruniitr commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,155 @@
+    for endpoint in endpoints:
+        meta_url = os.path.join(settings.DATA_HUB_METADATA_URL, endpoint)
+
+        print(f"Fetching {meta_url} ...")
```
updated with logging module functions

saruniitr commented:
Addressed review comments,
- Used django cache backend instead of using redis directly
- moved cache handling from periodic task to the fetch function
- basic unit tests as more tests is tricky in this case as explained in above comment. Appreciate any suggestions to improve this.

reupen commented on app/settings.py:
```diff
@@ -173,7 +173,46 @@
+    CELERY_RESULT_BACKEND = BROKER_URL
+    CELERY_ACCEPT_CONTENT = ['application/json']
+    CELERY_TASK_SERIALIZER = 'json'
+    CELERY_RESULT_SERIALIZER = 'json'
```
These are all the default values, there should be no reason to override them:
docs.celeryproject.org/en/latest/userguide/configuration.html

saruniitr replied:
these are included for completeness, they can be removed.

reupen commented on app/settings.py:
```diff
@@ -173,7 +173,46 @@
+HAWK_ID=env("HAWK_ID")
+HAWK_CREDENTIALS = {
+    env("HAWK_ID"): {
+        "id": env("HAWK_ID"),
```
Is storing `HAWK_ID` in three places needed?

saruniitr replied:
not really, I was trying group them an object but only ID and KEY are sufficient, will modify this.

reupen commented on app/enquiries/tasks.py:
```diff
@@ -0,0 +1,32 @@
+        logging.info(f"Data Hub metadata last refreshed at {datetime.now()}")
+    except Exception as e:
+        logging.error(f"Error refreshing metadata, {e}")
+        raise e
```
I'm not sure why `logging.error()` is needed – if the task fails, Celery will log the error, and we'll end up will double-Sentry alerts.

saruniitr replied:
My understanding is sentry only monitors exceptions? yes but if errors also then I can remove this.

reupen replied:
The logging integration is enabled by default in Sentry: https://docs.sentry.io/platforms/python/logging/

However, the Django and Celery integrations will probably need turning on (if they haven't been already).

saruniitr replied:
removed error log

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,204 @@
+    return dh_data[0]["id"] if dh_data else None
+
+
+def dh_company_search(company_name):
```
Is this meant to be in this PR?

saruniitr replied:
Yes, these are the main function in this PR. We will perform a company search and provide user with the matching results (assuming there are matches in DH). User can select the appropriate one and assign it to the enquiry. This gives an early indication/confidence to the user that enquiry is for a valid company and they can create a company if not found.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,204 @@
+    return companies
+
+
+def dh_contact_search(contact_name):
```
Is this meant to be in this PR?

saruniitr replied:
same as above

saruniitr commented:
Please let me know if there are any other comments

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,214 @@
+        if response.ok:
+            metadata[endpoint] = response.json()
+        else:
+            metadata["failed"].append(endpoint)
```
This will end up being cached – so what will happen when the page is displayed if some of the metadata requests failed?

saruniitr replied:
Yes but it is safe as it is only intended for debug purposes, no decisions are being taken based on this field. The code that actually uses these changes is not using this field. Once these changes are merged I can actually use these functions and see if it can be removed. For now I will retain this.

reupen commented:
> Yes, besides fetching metadata I wanted to include few other helper functions. I could've split but many other branches are dependent on this PR getting merged, it will further push review/merge of other branches.

If this is blocking other things I would suggest splitting things out. For example, company and contact search don't need to be part of this PR. As for metadata, I think the first step would be to add metadata retrieval without caching and add the caching afterwards.

saruniitr commented on app/enquiries/celery.py:
```diff
@@ -0,0 +1,12 @@
+# Using a string here means the worker will not have to
+# pickle the object when using Windows.
+app.config_from_object('django.conf:settings')
+app.autodiscover_tasks(lambda: settings.INSTALLED_APPS)
```
well it is not really required here but we can provide a list of packages to `autodiscover_tasks`
https://docs.celeryproject.org/en/stable/reference/celery.html

saruniitr commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,214 @@
+        if not cached_metadata:
+            logging.info("Metadata expired in cache, fetching again ...")
+            cached_metadata = _dh_fetch_metadata()
+            cache.set(cache_key, json.dumps(cached_metadata), timeout=expiry_secs)
```
yes, but using the data structure is better here, modified it.

saruniitr commented on app/enquiries/tasks.py:
```diff
@@ -0,0 +1,31 @@
+        dh_metadata = dh_fetch_metadata(expiry_secs=expiry_secs)
+        logging.info(f"Data Hub metadata last refreshed at {datetime.now()}")
+    except Exception as e:
+        raise e
```
yes, now it can be removed

saruniitr commented on app/settings.py:
```diff
@@ -173,7 +173,37 @@
 
 STATIC_URL = '/static/'
 
-
```
enabled integrations

saruniitr commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,214 @@
+
+    # It is not an error for us if the request fails, this can happen if the
+    # Access token is invalid, consider that there are no matches
+    if response.status_code != status.HTTP_200_OK:
```
done

saruniitr commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -0,0 +1,214 @@
+    response = dh_request("POST", url, payload)
+
+    # It is not an error for us if the request fails, this can happen if the
+    # Access token is invalid, consider that there are no matches
```
This is to specifically handle the case if access token is not valid then we get authentication credentials not provided error. This is an error wrt Data Hub api but not a fatal one for the application. We can assume it is same not finding any matches but the error information is propagated upwards (which gets shown to user). Perhaps the comment is not clear.
Company search is done when user edits enquiry to see if there are any matches. If there is a problem at this point we won't get results but user is notified of the error and can take action.

This won't happen once SSO is integrated as user will be automatically redirected to login if token expires, since it is not merged yet I am handling this here.

saruniitr commented:
I think it is better not to split at this stage because all of these functions are required for next PR. Only fetching metadata or company/contact search is not sufficient. Splitting now pushes things again. I think the tests situation has improved and will extend them further in the next PR.

reupen commented on app/enquiries/tasks.py:
```diff
@@ -0,0 +1,30 @@
+
+@periodic_task(
+    run_every=(crontab(minute="*/1")),
+    # run_every=(crontab(hour=FETCH_INTERVAL_HOURS, minute="0")),
```
Could you check this – wasn't the commented out line correct?

saruniitr replied:
yes commented line is correct, it was done for debug but forgot to remove while pushing these changes. I will correct this.

saruniitr commented:
thanks for the approval and I agree with your comment on metadata caching, I will be adding more tests in that area in next PR where investment creation feature uses this metadata.

saruniitr commented:
Thanks for the reviews, received two approvals, merging it.
